File: gamehub-lite-src/extension/CacheManager.java
package com.gamehub.lite.extension;

import android.util.Log;
import java.io.File;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class CacheManager {
    private static final String TAG = "GameHubNexus-Cache";
    private final File cacheDir;
    private final long maxCacheSize;

    public CacheManager(File cacheDir) {
        // Default 10 GB limit for local cached files
        this(cacheDir, 10L * 1024 * 1024 * 1024);
    }

    public CacheManager(File cacheDir, long maxCacheSize) {
        this.cacheDir = cacheDir;
        this.maxCacheSize = maxCacheSize;
        if (!cacheDir.exists()) {
            cacheDir.mkdirs();
        }
        enforceCacheLimit();
    }

    public long getCacheSize() {
        return getDirSize(cacheDir);
    }

    public void cleanAllCache() {
        Log.i(TAG, "Cleaning entire L2 cache...");
        File[] files = cacheDir.listFiles();
        if (files != null) {
            for (File file : files) {
                deleteRecursive(file);
            }
        }
    }

    public synchronized void enforceCacheLimit() {
        long currentSize = getCacheSize();
        Log.d(TAG, "Current L2 Cache Size: " + (currentSize / (1024 * 1024)) + " MB / " + (maxCacheSize / (1024 * 1024)) + " MB");

        if (currentSize <= maxCacheSize) {
            return;
        }

        Log.i(TAG, "Cache size limit exceeded. Evicting oldest files...");

        List<File> allFiles = new ArrayList<>();
        gatherFiles(cacheDir, allFiles);

        // Sort files by last modified time (oldest first)
        Collections.sort(allFiles, new Comparator<File>() {
            @Override
            public int compare(File f1, File f2) {
                return Long.compare(f1.lastModified(), f2.lastModified());
            }
        });

        long sizeToRemove = currentSize - maxCacheSize;
        for (File file : allFiles) {
            if (sizeToRemove <= 0) {
                break;
            }
            if (file.isFile()) {
                long fileSize = file.length();
                if (file.delete()) {
                    sizeToRemove -= fileSize;
                    Log.d(TAG, "Evicted cached file: " + file.getName() + " (Size: " + (fileSize / 1024) + " KB)");
                }
            }
        }

        Log.i(TAG, "Cache eviction completed. Current size: " + (getCacheSize() / (1024 * 1024)) + " MB");
    }

    private long getDirSize(File dir) {
        long size = 0;
        File[] files = dir.listFiles();
        if (files == null) return 0;
        for (File file : files) {
            if (file.isDirectory()) {
                size += getDirSize(file);
            } else {
                size += file.length();
            }
        }
        return size;
    }

    private void gatherFiles(File dir, List<File> list) {
        File[] files = dir.listFiles();
        if (files == null) return;
        for (File file : files) {
            if (file.isDirectory()) {
                gatherFiles(file, list);
            } else {
                list.add(file);
            }
        }
    }

    private void deleteRecursive(File file) {
        if (file.isDirectory()) {
            File[] children = file.listFiles();
            if (children != null) {
                for (File child : children) {
                    deleteRecursive(child);
                }
            }
        }
        file.delete();
    }
}

File: compiler-helper/src/main/kotlin/com/gamehub/lite/extension/CacheManager.kt
package com.gamehub.lite.extension

import android.util.Log
import java.io.File

class CacheManager(
    private val cacheDir: File,
    private val maxCacheSize: Long = 10L * 1024 * 1024 * 1024 // 10 GB limit for local cached files
) {
    private val TAG = "GameHubNexus-Cache"

    init {
        if (!cacheDir.exists()) {
            cacheDir.mkdirs()
        }
        enforceCacheLimit()
    }

    fun getCacheSize(): Long {
        return getDirSize(cacheDir)
    }

    fun cleanAllCache() {
        Log.i(TAG, "Cleaning entire L2 cache...")
        cacheDir.listFiles()?.forEach { file ->
            deleteRecursive(file)
        }
    }

    /**
     * Enforces the local cache size limit using Least Recently Used (LRU) policy.
     * Deletes the oldest files if cache directory exceeds maxCacheSize.
     */
    @Synchronized
    fun enforceCacheLimit() {
        val currentSize = getCacheSize()
        Log.d(TAG, "Current L2 Cache Size: ${currentSize / (1024 * 1024)} MB / ${maxCacheSize / (1024 * 1024)} MB")

        if (currentSize <= maxCacheSize) return

        Log.i(TAG, "Cache size limit exceeded. Evicting oldest files...")

        // Gather all files in cache recursively
        val allFiles = mutableListOf<File>()
        gatherFiles(cacheDir, allFiles)

        // Sort files by last modified time (oldest first)
        allFiles.sortBy { it.lastModified() }

        var sizeToRemove = currentSize - maxCacheSize
        for (file in allFiles) {
            if (sizeToRemove <= 0) break
            
            if (file.isFile) {
                val fileSize = file.length()
                if (file.delete()) {
                    sizeToRemove -= fileSize
                    Log.d(TAG, "Evicted cached file: ${file.name} (Size: ${fileSize / 1024} KB)")
                }
            }
        }
        
        Log.i(TAG, "Cache eviction completed. Current size: ${getCacheSize() / (1024 * 1024)} MB")
    }

    private fun getDirSize(dir: File): Long {
        var size: Long = 0
        val files = dir.listFiles() ?: return 0
        for (file in files) {
            size += if (file.isDirectory) {
                getDirSize(file)
            } else {
                file.length()
            }
        }
        return size
    }

    private fun gatherFiles(dir: File, list: MutableList<File>) {
        val files = dir.listFiles() ?: return
        for (file in files) {
            if (file.isDirectory) {
                gatherFiles(file, list)
            } else {
                list.add(file)
            }
        }
    }

    private fun deleteRecursive(file: File) {
        if (file.isDirectory) {
            file.listFiles()?.forEach { deleteRecursive(it) }
        }
        file.delete()
    }
}
